lib/ace/transport_app.rb

```diff
@@ -215,37 +217,32 @@ def self.trusted_facts(certname)
       # workflow, errors within the Configuer and not as pleasant - can have
       # _some_ control over them, especially around status codes from
       # the /v4/catalog endpoint
-      @plugins.with_synced_libdir(environment, certname) do
-        ACE::TransportApp.init_puppet_target(certname, body['target']['remote-transport'], body['target'])
-        configurer = ACE::Configurer.new(body['compiler']['transaction_uuid'], body['compiler']['job_id'])
-        configurer.run(transport_name: certname,
-                       environment: environment,
-                       network_device: true,
-                       pluginsync: false,
-                       trusted_facts: ACE::TransportApp.trusted_facts(certname))
-      end
-
-      # simulate expected error cases
-      if body['compiler']['certname'] == 'fail.example.net'
-        [200, { _error: {
-          msg: 'catalog compile failed',
-          kind: 'puppetlabs/ace/compile_failed',
-          details: 'upstream api errors go here'
-        } }.to_json]
-      elsif body['compiler']['certname'] == 'credentials.example.net'
-        [200, { _error: {
-          msg: 'target specification invalid',
-          kind: 'puppetlabs/ace/target_spec',
-          details: 'upstream api errors go here'
-        } }.to_json]
-      elsif body['compiler']['certname'] == 'reports.example.net'
-        [200, { _error: {
-          msg: 'report submission failed',
-          kind: 'puppetlabs/ace/reporting_failed',
-          details: 'upstream api errors go here'
-        } }.to_json]
+      begin
+        @plugins.with_synced_libdir(environment, certname) do
+          ACE::TransportApp.init_puppet_target(certname, body['target']['remote-transport'], body['target'])
+          configurer = ACE::Configurer.new(body['compiler']['transaction_uuid'], body['compiler']['job_id'])
+          configurer.run(transport_name: certname,
+                         environment: environment,
+                         network_device: true,
+                         pluginsync: false,
+                         trusted_facts: ACE::TransportApp.trusted_facts(certname))
+        end
+      rescue ACE::Error => e
+        process_error = {  _error: e.to_h }
+        return [400, process_error.to_json]
+      rescue StandardError => e
+        process_error = { _error: ACE::Error.new(e.message,
+                                                 'puppetlabs/ace/processing_exception',
+                                                 class: e.class, backtrace: e.backtrace) }
+        return [400, process_error.to_json]
```

Comment:
At this point a standard error I do not think could be classed as a `400 Bad Request` as the request body has already been validated. A `StandardError` I guess is anything that could potentially bubble up from the `configurer`,  which means it would more likely fall in to a server error vs a client error.

Reply:
Yeah, great point. I'll switch this out to a 500